pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go

```diff
@@ -158,21 +179,137 @@ func (dswp *desiredStateOfWorldPopulator) Run(ctx context.Context, sourcesReady
 		dswp.hasAddedPods = true
 	}
 	dswp.hasAddedPodsLock.Unlock()
-	wait.UntilWithContext(ctx, dswp.populatorLoop, dswp.loopSleepDuration)
+
+	// initially the loopSleepDuration is set to 100ms (min period)
+	// the only check needed is wether the FG is enabled
+	if utilfeature.DefaultFeatureGate.Enabled(features.EventedDesiredStateOfWorldPopulator) {
+		dswp.setWatchProcessStoped(false)
+		go func() {
+			_ = wait.PollUntilContextCancel(ctx, 0, false, func(ctx context.Context) (bool, error) {
+				return dswp.watchAndProcessEvent(ctx), nil
+			})
+		}()
+	}
+	wait.DyanmicPeriodUntil(ctx, func() {
+		dswp.populatorLoop(ctx)
+		if utilfeature.DefaultFeatureGate.Enabled(features.EventedDesiredStateOfWorldPopulator) {
+			if !dswp.isWatchProcessStoped() {
+				dswp.increaseLoopSleepDuration(ctx)
+			} else if dswp.getLoopSleepDuration() != initialLoopSleepDuration {
+				// make sure after the watch func is stoped => reset the loopSleepDuration to the initial default value
+				dswp.setLoopSleepDuration(100 * time.Millisecond)
+			}
+		}
+	}, func() time.Duration {
+		return dswp.getLoopSleepDuration()
+	}, 0, true)
+}
+
+func (dswp *desiredStateOfWorldPopulator) increaseLoopSleepDuration(ctx context.Context) {
+	if dswp.executionCounter > dswpEventMinLoopBeforeIncreasingSleepPeriod && dswp.getLoopSleepDuration() == initialLoopSleepDuration {
+		dswp.executionCounter = 0
+	}
+	dswp.executionCounter++
+	// gradually increase after the third execution
+	// ( +100ms on each iteration) the sleep period to a maximum ( 1 second)
+	if dswp.getLoopSleepDuration() < initialLoopSleepDuration*10 && dswp.executionCounter > dswpEventMinLoopBeforeIncreasingSleepPeriod {
+		dswp.setLoopSleepDuration(dswp.getLoopSleepDuration() + 100*time.Millisecond)
+	}
+}
+
+// Subscribe to CRI event stream and fallback to 100ms loopSleepDuration
+// when CRI runtime returns errors
+func (dswp *desiredStateOfWorldPopulator) watchAndProcessEvent(ctx context.Context) bool {
+	logger := klog.FromContext(ctx)
+	logger.Info("Evented DSWP: watchAndProcessEvent starts")
+
+	containerEventsResponseCh := make(chan *runtimeapi.ContainerEventResponse, dswpEventChannelCapacity)
+	defer close(containerEventsResponseCh)
+
+	go func() {
+		numAttempts := 0
+		for {
+			if numAttempts >= dswpEventMaxStreamRetries {
+				logger.Info("Evented DSWP: Disable subscription to CRI event")
+				dswp.setLoopSleepDuration(initialLoopSleepDuration)
+				logger.Info("Evented DSWP: watchAndProcessEvent exits")
+				dswp.setWatchProcessStoped(true)
+				break
+			}
+
+			// Get the container events from the runtime.
+			err := dswp.runtimeService.GetContainerEvents(context.Background(), containerEventsResponseCh, func(runtimeapi.RuntimeService_GetContainerEventsClient) {
```

Comment:
@HirazawaUi , how this feature retrieves CRI  events  .

Reply:
FYI : The current PR needs to be updated with the latest [kep changes](https://github.com/kubernetes/enhancements/pull/4980).